text/0000-ubi-based-stacks.md

```diff
@@ -0,0 +1,181 @@
+within environments with existing UBI deployments. In addition, use of a UBI image can sometimes be a requirement 
+for an application qualifying for support from some vendors while in production. That requirement is also extended 
+to any runtime in use, eg for Node or Java, the runtime must be installed via rpm/dnf, rather than via a tarball. 
+
+## Detailed Explanation
+
+### Background
+
+There are a wide variety of UBI images offered by Red Hat, offering pre configured deployments of various runtimes 
+(such as Java), and offering cutdown images designed for extension (ubi-minimal).
+
+While it would be possible to use UBI images in a similar fashion to the existing Ubuntu based images, and have 
+buildpacks perform the download & install of depdencies within layers, the preferred route would be to to use the 
+dependencies supplied via the configured package repositories within the images. Eg, rather than downloading a 
+JVM/JDK, one would be installed via 'yum' or 'dnf'. 
+
+Until recently, CNCF Buildpacks were unable to effect changes to the Builder or Run image at a system level (eg, you
+could not make modifications that required root from a buildpack, such as installing packages via yum). Recent 
+updates to the Buildpacks project have added 'extensions', these can be considered as special variants of buildpacks 
+that are able to modify the Builder image (and substitute the Run image), via generated Dockerfiles created during 
+the execution of the Extensions. (In the future, the goal is to also allow modification of the Run image via the 
+same mechanism).
+
+Via extensions, it becomes possible to have a Builder image that can install dependencies such as Java, Maven,
+NodeJS, etc via yum. These extensions are very similar to buildpacks, participating in the 'detect' phase, and 
+being allowed to issue 'provides' to the buildplan (but critically not 'requires').. then extensions that are
+selected to become part of a build run a 'generate' task (rather than a buildpacks 'build' task), that may output 
+Dockerfiles to modify the Builder image, and/or substitute the Run image. Just like buildpacks, the detect & 
+generate invocations get access to the source project to determine what actions they wish to take. For more 
+information on extensions see.. https://github.com/buildpacks/spec/blob/main/image_extension.md
+
+### Initial offered image
+
+The intent is provide an initial builder image that caters to Java & NodeJS runtimes (without Java Native). 
+Some early prototyping has shown this to be possible using the existing Java & NodeJS builder ecosystems 
+present within Paketo today, with Extensions installing the main runtime dependencies. 
+
+In the case of Java, much work has already been done by Paketo to allow for different buildpacks to supply the 
+JDK/JVM, making it relatively simple to extend this to allow an Extension to perform the same role. For NodeJS, there
```

Comment:
Are any changes to the Java buildpack required to support this? Or will it Just Work™?

Reply:
Tested and working within local prototypes with no changes required so far =). 

The flow is basically.. extensions & buildpacks run their detects, both extension-java and buildpack-supplying-java are selected.. (both also claim to 'provides' jdk & jre) the extension runs first, installs java via rpm, then when the buildpack-supplying-java executes the build phase it notices the buildplan no longer requires jdk/jre and becomes a no-op. 
 
It's also possible to just omit the jdk/jre supplying buildpacks from the ubi8 builder image, but I'm trying here to leave as much of the existing meta buildpack content as-is as possible. 
 
 For what it's worth, Node did need a minor change to give it the same behavior, as it wasn't correctly handling the removal of the requires from the buildplan. I believe that change has already been contributed & released.

Reply:
That's great. Thanks for digging into that and clarifying.